server/src/test/java/io/crate/planner/operators/JoinTest.java

```diff
@@ -1142,4 +1142,88 @@ public void test_right_outer_join_rewrite_to_left_outer_join() throws Exception
+            "NestedLoopJoin[CROSS]",
+            "  ├ NestedLoopJoin[CROSS]",
+            "  │  ├ NestedLoopJoin[CROSS]",
+            "  │  │  ├ NestedLoopJoin[CROSS]",
+            "  │  │  │  ├ NestedLoopJoin[RIGHT | true]",
+            "  │  │  │  │  ├ Collect[doc.t1 | [a] | true]",
+            "  │  │  │  │  └ Collect[doc.t2 | [b] | true]",
+            "  │  │  │  └ Collect[doc.t3 | [c] | true]",
+            "  │  │  └ Collect[doc.t4 | [d] | true]",
+            "  │  └ Collect[doc.t5 | [e] | true]",
+            "  └ Collect[doc.t6 | [f] | true]"
+        );
+
+
+        mss = e.analyze("SELECT * FROM doc.t1, doc.t2, doc.t3 right join doc.t4 on true, doc.t5, doc.t6");
+        result = buildLogicalPlan(mss, plannerCtx);
+
+        assertThat(result).hasOperators(
+            "Eval[a, b, c, d, e, f]",
+            "  └ NestedLoopJoin[CROSS]",
+            "    ├ NestedLoopJoin[CROSS]",
+            "    │  ├ NestedLoopJoin[CROSS]",
+            "    │  │  ├ NestedLoopJoin[CROSS]",
+            "    │  │  │  ├ NestedLoopJoin[RIGHT | true]",
+            "    │  │  │  │  ├ Collect[doc.t3 | [c] | true]",
+            "    │  │  │  │  └ Collect[doc.t4 | [d] | true]",
+            "    │  │  │  └ Collect[doc.t1 | [a] | true]",
+            "    │  │  └ Collect[doc.t2 | [b] | true]",
+            "    │  └ Collect[doc.t5 | [e] | true]",
+            "    └ Collect[doc.t6 | [f] | true]"
+        );
+
+        mss = e.analyze("SELECT * FROM doc.t1, doc.t2 right join doc.t3 on true, doc.t4 left join doc.t5 on true, doc.t6");
+        result = buildLogicalPlan(mss, plannerCtx);
+
+        assertThat(result).hasOperators(
+            "Eval[a, b, c, d, e, f]",
+            "  └ NestedLoopJoin[CROSS]",
+            "    ├ NestedLoopJoin[LEFT | true]",
+            "    │  ├ NestedLoopJoin[CROSS]",
+            "    │  │  ├ NestedLoopJoin[CROSS]",
+            "    │  │  │  ├ NestedLoopJoin[RIGHT | true]",
+            "    │  │  │  │  ├ Collect[doc.t2 | [b] | true]",
+            "    │  │  │  │  └ Collect[doc.t3 | [c] | true]",
+            "    │  │  │  └ Collect[doc.t1 | [a] | true]",
+            "    │  │  └ Collect[doc.t4 | [d] | true]",
+            "    │  └ Collect[doc.t5 | [e] | true]",
+            "    └ Collect[doc.t6 | [f] | true]"
+        );
+
+        mss = e.analyze("SELECT * FROM doc.t1, doc.t2 right join doc.t3 on true left join doc.t5 on true, doc.t6");
+        result = buildLogicalPlan(mss, plannerCtx);
+
+        assertThat(result).hasOperators(
+            "Eval[a, b, c, e, f]",
+            "  └ NestedLoopJoin[CROSS]",
+            "    ├ NestedLoopJoin[LEFT | true]",
+            "    │  ├ NestedLoopJoin[CROSS]",
+            "    │  │  ├ NestedLoopJoin[RIGHT | true]",
+            "    │  │  │  ├ Collect[doc.t2 | [b] | true]",
+            "    │  │  │  └ Collect[doc.t3 | [c] | true]",
+            "    │  │  └ Collect[doc.t1 | [a] | true]",
+            "    │  └ Collect[doc.t5 | [e] | true]",
```

Comment:
Shouldn't we apply all outers first?

ie (t2 x t3) x t5

Here we join (t2 x t3) with t1 first

Reply:
The original test asserted the result aligned with Postgres.

Reply:
But let me see. All cross/inner-joins can be moved around the only important part is that right join is executed before the left-join which is still the case.

Reply:
As mentioned in the description the order is not preserved anymore. Anyway, this is a best-effort fix. The join-pair model together with the JoinPlanBuilder is perfect to process a tree-like structure which is deconstructed into pairs of nodes.

Reply:
One year i started to rewrite to whole join building, but we dropped it because there were no real issue at the time with that besides complexity. Maybe we reconsider that in the next cycle.

Reply:
I was not worried about original order but rather about the correctness

If it's aligned with Postgres, then fine. I somehow understood commit message that all outer joins must be applied first